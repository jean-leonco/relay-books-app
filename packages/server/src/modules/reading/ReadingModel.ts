import mongoose, { Document, Model, Types } from 'mongoose';

import { isActiveMongooseField, removedAtMongooseField } from '../../core/mongoose/withMongooseFields';

const { ObjectId } = mongoose.Schema.Types;

const Schema = new mongoose.Schema(
  {
    userId: {
      type: ObjectId,
      ref: 'User',
      description: 'The user that is reading.',
      required: true,
      index: true,
    },
    bookId: {
      type: ObjectId,
      ref: 'Book',
      description: 'The book being read.',
      required: true,
      index: true,
    },
    readPages: {
      type: Number,
      description: 'The total read pages. ex: 50',
      required: true,
    },
    ...isActiveMongooseField,
    ...removedAtMongooseField,
  },
  {
    collection: 'Reading',
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
  },
);

export interface IReading extends Document {
  userId: Types.ObjectId;
  bookId: Types.ObjectId;
  readPages: number;
  isActive: boolean;
  removedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const ReadingModel: Model<IReading> = mongoose.model('Reading', Schema);

export default ReadingModel;
