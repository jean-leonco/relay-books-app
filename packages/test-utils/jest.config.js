const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTestFramework.js'],
  globalSetup: '<rootDir>/test/setup.js',
  globalTeardown: '<rootDir>/test/teardown.js',
  transform: {
    // '^.+\\.(js|jsx|ts|tsx)?$': require.resolve('babel-jest'),
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/test/babel-transformer',
  },
};
