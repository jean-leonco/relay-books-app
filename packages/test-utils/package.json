{
  "name": "@booksapp/test-utils",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@booksapp/babel": "^0.0.1",
    "@booksapp/types": "^0.0.1",
    "@koa/router": "^9.3.1",
    "babel-jest": "^26.1.0",
    "graphql-relay": "0.6.0",
    "koa": "^2.13.0",
    "koa-bodyparser": "^4.3.0",
    "mongoose": "^5.10.7",
    "supertest": "^5.0.0"
  }
}
