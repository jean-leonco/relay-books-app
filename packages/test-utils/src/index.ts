export { defaultFrozenKeys, sanitizeValue, sanitizeTestObject } from './objectNormalization';
export { httpRequestGraphql } from './httpRequestGraphql';
