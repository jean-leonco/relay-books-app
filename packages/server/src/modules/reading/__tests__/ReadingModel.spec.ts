import { sanitizeTestObject } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createUser,
  createReading,
  createBook,
} from '../../../../test/helpers';

import ReadingModel, { IReading } from '../ReadingModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('ReadingModel', () => {
  it('should create new reading using ReadingModel', async () => {
    const user = await createUser();
    const book = await createBook({ pages: 10 });
    const readPages = 5;

    const reading = await new ReadingModel({
      userId: user._id,
      bookId: book._id,
      readPages,
    }).save();

    const readingObj = await ReadingModel.findOne({ _id: reading._id }).lean<IReading>();

    expect(readingObj?.userId).toMatchObject(user._id);
    expect(readingObj?.bookId).toMatchObject(book._id);
    expect(readingObj?.readPages).toBe(readPages);
    expect(sanitizeTestObject(readingObj)).toMatchSnapshot();
  });

  it('should create new reading using createRow', async () => {
    const user = await createUser();
    const book = await createBook({ pages: 10 });

    const reading = await createReading();

    const readingObj = await ReadingModel.findOne({ _id: reading._id }).lean<IReading>();

    expect(readingObj?.userId).toMatchObject(user._id);
    expect(readingObj?.bookId).toMatchObject(book._id);
    expect(readingObj?.readPages).toBe(1);
  });

  it('should create new reading with custom properties', async () => {
    const user = await createUser();
    const book = await createBook({ pages: 10 });
    const readPages = 5;

    const reading = await createReading({ userId: user._id, bookId: book._id, readPages });

    const readingObj = await ReadingModel.findOne({ _id: reading._id }).lean<IReading>();

    expect(readingObj?.userId).toMatchObject(user._id);
    expect(readingObj?.bookId).toMatchObject(book._id);
    expect(readingObj?.readPages).toBe(readPages);
  });
});
