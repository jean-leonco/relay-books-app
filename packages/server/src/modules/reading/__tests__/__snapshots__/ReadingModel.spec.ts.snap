// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingModel should create new reading using ReadingModel 1`] = `
Object {
  "__v": 0,
  "_id": "ObjectId",
  "bookId": "ObjectId",
  "createdAt": "FROZEN-CREATEDAT",
  "isActive": true,
  "readPages": 5,
  "removedAt": "EMPTY",
  "updatedAt": "FROZEN-UPDATEDAT",
  "userId": "ObjectId",
}
`;
